### ISSUE TITLE
Feature Request: creating a bunch of small panes of specific size (like 80x2) with split-window

### ISSUE THREAD

**Issue by cnst** [tmux/tmux#2789]
### Issue description

People often use tmux as terminal multiplexer and a window manager, and it may also be useful to use it in shell scripts for parallel processing of some commands.

An issue for using `tmux split-window` from within simple shell scripts is that you have to be able to fit all the new panes in a single window, and there's a very small number of panes you can create before running out of screen space.

For example, on a MacBook Pro 16", you may have a tmux window of size 254x67 (as per `tmux display-panes -d0`).  Ideally, the smallest usable pane size may be `80x2`, which requires roughly `81x3` each together with the tmux border.  For `X`, `254/81 = 3.13`; for `Y`, `67/3 = 22.33`; this means we could fit `3x22` panes of `84x2` each (`X=254/3=84.66`; `84*3+2=254` and `Y=67/22=3.04`), i.e., a total of `3*22 = 66` panes of `84x2` each in a single window.  Not bad!

However, the above requires doing some math and bookkeeping on user's part, and the simplest solution of creating vertical split of 2 lines each limits us to just 22 panes total of `254x2` each:

```
for COMMAND in $COMMANDS_TO_EXECUTE ; do tmux split-window -t.0 -l2 "$COMMAND ; read" ; done
```

The above could be modified to instead do a horizontal split x3 first, and then have 3 loops on each on the 3 original `84x67` panes to split each of the original 3 panes into 22 smaller 2-line panes, to get a total of `3x22` panes of size `84x2`, but it'd be nice if there was some easy way to automate this, maybe even in a way where overflow panes would automatically result in creation of an extra window, and the extra panes would be stuffed in the new window.

**Comment by nicm** [comment]
I'm not very clear exactly what you are asking for but it seems so specific that it would be best to script it once for your particular purpose.

If you want to create 22 panes and then create a new window, you can so something like:
~~~~
tmux new -d -x- -y-;n=0;for i in `seq 50`; do n=`expr $n + 1`; [ $n -eq 22 ] && tmux neww "echo $i; exec $SHELL" || tmux splitw "echo $i; exec $SHELL" \; selectl tiled; done
~~~~
Or if you want to create as many as you can vertically and make a new window when you run out of space, something like:
~~~~
tmux new -d -x- -y-;for i in `seq 50`; do tmux splitw -bf -l2 "echo $i; exec $SHELL" || tmux neww "echo $i; exec $SHELL"; done
~~~~

**Comment by nicm** [comment]
There are also existing scripts like https://github.com/greymd/tmux-xpanes you might want to look at.

**Comment by cnst** [comment]
tl;dr: I want to create as many panes as possible of size at least 80x2 in a 254x67 window, e.g., about 66 panes per math above.

Is that what tmux-xpanes does?  Thanks for the pointer, will take a look.

**Comment by nicm** [comment]
Making them specifically 80x2 would be harder than just generating rows of 3 which you can do pretty easily:
~~~~
tmux neww
while :; do
    tmux splitw -hl 66%
    tmux splitw -hl 50%
    tmux splitw -bf -l2 || break
done

**Comment by cnst** [comment]
Oh, nice, looks like `select-layout tiled` already does nearly what I want.  The documentation is not entirely clear how to define a minimum height or width, though.

Thanks for the other pointers as well!

**Comment by nicm** [comment]
`tiled` you cannot specify a minimum width or height, you just get what you are given, the builtin layouts are pretty basic and not very configurable. If you create 66 panes and then do `selectl tiled` it will probably make each pane 31x8.

I just live with that or use `even-vertical` myself (but that would only give you 30 or so panes).

Preset layouts were never developed much further since they were written, it would be nice if they were more configurable and if there were a few more, especially for larger monitors.

But is is generally pretty easy to script layouts anyway.

**Comment by cnst** [comment]
I've tried your `splitw -bf -l2` suggestion above, but only the top 3 panes were the specified number of lines, the rest were just one line, is there a way to have all panes be a minimum size of `-l2`?

**Comment by nicm** [comment]
Yes... `splitw -f` will resize other panes in a way that isn't ideal. I think you will need to work out how many rows first which is not as elegant, something like this maybe:
~~~~
tmux neww
rows=$(tmux display -p '#{e|-:#{e|/:#{window_height},3},1}')
for i in `seq $rows`; do
    tmux splitw -l2
    tmux splitw -hl 66%
    tmux splitw -hl 50%
    tmux selectp -t'{top}'
done
tmux splitw -hl 66%
tmux splitw -hl 50%

**Comment by nicm** [comment]
Or this avoids precalculating, but if you have an even number of lines in the terminal you will end up with two 1-line panes:
~~~~
tmux neww
while :; do
    tmux splitw -dl2 || break
done
while :; do
    tmux splitw -hl66% \; splitw -hl50%
    [ $(tmux display -p '#{pane_at_bottom}')  -eq 1 ] && break
    tmux selectp -t{down-of}
done
~~~~
Anyway I'm sure you get the idea.

**Comment by cnst** [comment]
Thanks for so many hints.  BTW, I've tried using some of these scripts together with `sleep`, and the commands have interfered with the panes in other windows in the session.

Shouldn't the commands apply to the window from which they execute?  Or they always intended to apply to the one in focus?

**Comment by nicm** [comment]
You will need to use targets to avoid this, for example when you create the window, save its target:
~~~~
W=$(tmux neww -P)
~~~~
Or pass the ID into the script when you run it from a key binding:
~~~~
bind q run "sh /path/to/my/script '#{window_id}'"
~~~~
Then when you do all the other commands in the script, give them the target window:
~~~~
tmux splitw -t$W
~~~~
